pkg/tinydtls/contrib/sock_dtls.c

```diff
@@ -217,22 +306,61 @@ static int _get_ecdsa_key(struct dtls_context_t *ctx, const session_t *session,
                           const dtls_ecdsa_key_t **result)
 {
     (void)session;
-    int ret;
+    int ret = CREDMAN_ERROR;
     sock_dtls_t *sock = (sock_dtls_t *)dtls_get_app_data(ctx);
+    sock_udp_ep_t ep;
+
+    _session_to_ep(session, &ep);
 
     credman_credential_t credential;
-    ret = credman_get(&credential, sock->tag, CREDMAN_TYPE_ECDSA);
-    if (ret < 0) {
-            DEBUG("sock_dtls: no matching ecdsa credential found\n");
+    credential.tag = CREDMAN_TAG_EMPTY;
+    DEBUG("sock_dtls: get ECDSA key\n");
+
+    /* if the application set a callback , try to select credential from there */
+    if (sock->rpk_cb) {
+        DEBUG("sock_dtls: requesting the application\n");
+        credential.tag = sock->rpk_cb(sock, &ep, sock->tags, sock->tags_len);
+        if (credential.tag != CREDMAN_TAG_EMPTY) {
+            ret = credman_get(&credential, credential.tag, CREDMAN_TYPE_ECDSA);
+            if (ret != CREDMAN_OK) {
+                credential.tag = CREDMAN_TAG_EMPTY;
+            }
+        }
+    }
+
+    if (credential.tag == CREDMAN_TYPE_EMPTY) {
+        /* if could not get credential try to fetch the first valid credential */
+        for (unsigned i = 0; i < sock->tags_len; i++) {
+            ret = credman_get(&credential, sock->tags[i], CREDMAN_TYPE_ECDSA);
+            if (ret == CREDMAN_OK) {
+                break;
+            }
+        }
+
+        if (ret != CREDMAN_OK) {
+            DEBUG("sock_dtls: no valid credential registered\n");
             return dtls_alert_fatal_create(DTLS_ALERT_INTERNAL_ERROR);
+        }
+    }
+
+    ecdsa_key_assignment_t *key = NULL;
+    for (unsigned i = 0; i < CONFIG_DTLS_CREDENTIALS_MAX; i++) {
+        if (!_ecdsa_keys[i].session) {
+            key = &_ecdsa_keys[i];
```

Comment:
Please document this. You assign the key if there is a free session "slot"?

Reply:
Yes, tinyDTLS requires that the structure holding the key is allocated by the callback, so I introduced a pool of them.

Reply:
Comment added